### PR TITLE
Add CIFuzz external entrypoint for cifuzz base

### PR DESCRIPTION
**Pull request by mpherman2**
This PR will let Prow run CIFuzz by using the CIFuzz base image

### CODE REVIEW

File: infra/cifuzz/cifuzz_external_entrypoint.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
+# Copyright 2020 Google LLC
```
Comment (jonathanmetzman) [inline infra/cifuzz/cifuzz_external_entrypoint.py:1]: nit: Can you rename this file `cifuzz_combined_entrypoint`. I'm not sure it's going to be the preferred method for users that can't build two docker images.

File: infra/cifuzz/run_fuzzers_entrypoint.py [hunk 2]
```diff
@@ -42,6 +42,36 @@ def delete_unneeded_docker_images(config):
   docker.delete_images(images)
 
 
+def run_fuzzers_entry():
+  """ This is the entrypoint for the run_fuzzers github action.
+  This action can be added to any OSS-Fuzz project's workflow that uses
```
Comment (jonathanmetzman) [inline infra/cifuzz/run_fuzzers_entrypoint.py:47]: Can you delete this sentence please.

**Review by jonathanmetzman — COMMENTED** [review]
Hey this mostly looks OK. Except the comments aren't great (99% of that is not your fault) so I left some requests to improve them.
Are you sure this is needed for prow by the way?

File: infra/cifuzz/build_fuzzers_entrypoint.py [hunk 3]
```diff
@@ -28,30 +28,11 @@
     level=logging.DEBUG)
 
 
-def main():
+def build_fuzzers_entry():
```
[the remark this answers is not in the document]
Comment (mpherman2) [reply infra/cifuzz/build_fuzzers_entrypoint.py]: I assume you meant build_fuzzers_entrypoint()

File: infra/cifuzz/build_fuzzers_entrypoint.py [hunk 4]
```diff
@@ -28,30 +28,11 @@
...
   """Build OSS-Fuzz project's fuzzers for CI tools.
```
[the remark this answers is not in the document]
Comment (mpherman2) [reply infra/cifuzz/build_fuzzers_entrypoint.py]: Done

File: infra/cifuzz/build_fuzzers_entrypoint.py [hunk 5]
```diff
@@ -90,5 +71,32 @@ def main():
   return returncode
 
 
+def main():
+  """Build OSS-Fuzz project's fuzzers for CI tools.
```
[the remark this answers is not in the document]
Comment (mpherman2) [reply infra/cifuzz/build_fuzzers_entrypoint.py]: Done

File: infra/cifuzz/build_fuzzers_entrypoint.py [hunk 6]
```diff
@@ -90,5 +71,32 @@ def main():
...
+  This script is used to kick off the Github Actions CI tool. It is the
+  entrypoint of the Dockerfile in this directory. This action can be added to
```
[the remark this answers is not in the document]
Comment (mpherman2) [reply infra/cifuzz/build_fuzzers_entrypoint.py]: Done

File: infra/cifuzz/build_fuzzers_entrypoint.py [hunk 7]
```diff
@@ -90,5 +71,32 @@ def main():
...
+  any OSS-Fuzz project's workflow that uses Github.
+
+  Note: The resulting clusterfuzz binaries of this build are placed in
```
[the remark this answers is not in the document]
Comment (mpherman2) [reply infra/cifuzz/build_fuzzers_entrypoint.py]: Done

File: infra/cifuzz/build_fuzzers_entrypoint.py [hunk 8]
```diff
@@ -90,5 +71,32 @@ def main():
...
+    GITHUB_EVENT_NAME: The name of the hook event that triggered this script.
+    GITHUB_EVENT_PATH:
+      The path to the file containing the POST payload of the webhook:
+      https://help.github.com/en/actions/reference/virtual-environments-for-github-hosted-runners#filesystems-on-github-hosted-runners
+    GITHUB_WORKSPACE: The shared volume directory where input artifacts are.
+    DRY_RUN: If true, no failures will surface.
+    SANITIZER: The sanitizer to use when running fuzzers.
+
+  Returns:
+    0 on success or 1 on failure.
```
[the remark this answers is not in the document]
Comment (mpherman2) [reply infra/cifuzz/build_fuzzers_entrypoint.py]: Done. And changed CIfuzz_external_entrypoint logic to account for this as well.

File: infra/cifuzz/cifuzz_external_entrypoint.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mpherman2) [reply infra/cifuzz/cifuzz_external_entrypoint.py]: done

Comment (mpherman2) [reply infra/cifuzz/cifuzz_external_entrypoint.py]: done

File: infra/cifuzz/cifuzz_external_entrypoint.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Builds fuzzers and runs fuzzers. Entrypoint used for external users"""
+import sys
+
+import build_fuzzers_entrypoint
+import run_fuzzers_entrypoint
+
+# pylint: disable=c-extension-no-member
```
[the remark this answers is not in the document]
Comment (mpherman2) [reply infra/cifuzz/cifuzz_external_entrypoint.py]: Done

File: infra/cifuzz/cifuzz_external_entrypoint.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+# pylint gets confused because of the relative import of cifuzz.
+
+
+def main():
+  """Build OSS-Fuzz project's fuzzers for CI tools.
```
[the remark this answers is not in the document]
Comment (mpherman2) [reply infra/cifuzz/cifuzz_external_entrypoint.py]: Done

File: infra/cifuzz/cifuzz_external_entrypoint.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+  This script is used to kick off the Github Actions CI tool. It is the
+  entrypoint of the Dockerfile in this directory. This action can be added to
+  any OSS-Fuzz project's workflow that uses Github.
+
+  NOTE: The resulting clusterfuzz binaries of this build are placed in
```
[the remark this answers is not in the document]
Comment (mpherman2) [reply infra/cifuzz/cifuzz_external_entrypoint.py]: Done

File: infra/cifuzz/cifuzz_external_entrypoint.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+  the directory: ${GITHUB_WORKSPACE}/out
+
+  NOTE: libFuzzer binaries must be located in the ${GITHUB_WORKSPACE}/out
```
[the remark this answers is not in the document]
Comment (mpherman2) [reply infra/cifuzz/cifuzz_external_entrypoint.py]: Done

File: infra/cifuzz/cifuzz_external_entrypoint.py [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+  directory in order for this action to be used. This action will only fuzz the
+  binaries that are located in that directory. It is recommended that you add
+  the build_fuzzers action preceding this one.
+
+  NOTE: Any crash report will be in the filepath:
+  ${GITHUB_WORKSPACE}/out/testcase
+  This can be used in parallel with the upload-artifact action to surface the
```
[the remark this answers is not in the document]
Comment (mpherman2) [reply infra/cifuzz/cifuzz_external_entrypoint.py]: Done

File: infra/cifuzz/cifuzz_external_entrypoint.py [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+      https://help.github.com/en/actions/reference/virtual-environments-for-github-hosted-runners#filesystems-on-github-hosted-runners
+    GITHUB_WORKSPACE: The shared volume directory where input artifacts are.
+    DRY_RUN: If true, no failures will surface.
+    SANITIZER: The sanitizer to use when running fuzzers.
+    FUZZ_SECONDS: The length of time in seconds that fuzzers are to be run.
+
+  Returns:
+    0 on success or 1 on failure.
+  """
+  if build_fuzzers_entrypoint.build_fuzzers_entry() == 1:
```
[the remark this answers is not in the document]
Comment (mpherman2) [reply infra/cifuzz/cifuzz_external_entrypoint.py]: Done

[the remark this answers is not in the document]
Comment (mpherman2) [reply infra/cifuzz/cifuzz_external_entrypoint.py]: Done

File: infra/cifuzz/run_fuzzers_entrypoint.py [hunk 15]
```diff
@@ -42,6 +42,36 @@ def delete_unneeded_docker_images(config):
   docker.delete_images(images)
 
 
+def run_fuzzers_entry():
```
[the remark this answers is not in the document]
Comment (mpherman2) [reply infra/cifuzz/run_fuzzers_entrypoint.py]: Done

File: infra/cifuzz/run_fuzzers_entrypoint.py [hunk 16]
```diff
@@ -42,6 +42,36 @@ def delete_unneeded_docker_images(config):
...
+  """ This is the entrypoint for the run_fuzzers github action.
```
[the remark this answers is not in the document]
Comment (mpherman2) [reply infra/cifuzz/run_fuzzers_entrypoint.py]: Done

[the remark this answers is not in the document]
Comment (mpherman2) [reply infra/cifuzz/run_fuzzers_entrypoint.py]: Done

File: infra/cifuzz/run_fuzzers_entrypoint.py (hunk 2 above)
Comment (mpherman2) [reply infra/cifuzz/run_fuzzers_entrypoint.py:47]: Done

File: infra/cifuzz/run_fuzzers_entrypoint.py [hunk 17]
```diff
@@ -67,30 +97,7 @@ def main():
   Returns:
     0 on success or 1 on failure.
```
[the remark this answers is not in the document]
Comment (mpherman2) [reply infra/cifuzz/run_fuzzers_entrypoint.py]: Done

**Review by mpherman2 — COMMENTED** [review]
Sorry for the delay. Was working on a Hackathon last week. This is needed as Prow can't run two images in sequence. This makes it substantially easier to run this as a prow job (no need to write an extra script and bother with docker in docker). Alternatively we could make a new image if you do not wish cifuzz_base to have an entrypoint.

File: infra/cifuzz/cifuzz-base/Dockerfile [hunk 18]
```diff
@@ -35,4 +35,9 @@ RUN ldconfig
 ENV OSS_FUZZ_ROOT=/opt/oss-fuzz
 ADD . ${OSS_FUZZ_ROOT}
 RUN python3 -m pip install -r ${OSS_FUZZ_ROOT}/infra/cifuzz/requirements.txt
-RUN rm -rf ${OSS_FUZZ_ROOT}/infra
\ No newline at end of file
+RUN rm -rf ${OSS_FUZZ_ROOT}/infra
+
+# Python file to execute when the docker container starts up
```
Comment (jonathanmetzman) [inline infra/cifuzz/cifuzz-base/Dockerfile:40]: nit: end this line with a period please.

File: infra/cifuzz/cifuzz_combined_entrypoint.py [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+    GITHUB_WORKSPACE: The shared volume directory where input artifacts are.
+    DRY_RUN: If true, no failures will surface.
+    SANITIZER: The sanitizer to use when running fuzzers.
+    FUZZ_SECONDS: The length of time in seconds that fuzzers are to be run.
+
+  Returns:
+    0 on success or 1 on failure.
+  """
+  logging.debug("Using cifuzz_combined_entrypoint.")
+  build = build_fuzzers_entrypoint.build_fuzzers_entry()
```
Comment (jonathanmetzman) [inline infra/cifuzz/cifuzz_combined_entrypoint.py:46]: Shouldn't this be build_fuzzers_entrypoint?

File: infra/cifuzz/cifuzz_combined_entrypoint.py [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+  if build != 0:
+    return build
+  return run_fuzzers_entrypoint.run_fuzzers_entry()
```
Comment (jonathanmetzman) [inline infra/cifuzz/cifuzz_combined_entrypoint.py:49]: Shouldn't this be run_fuzzers_entrypoint?

File: infra/cifuzz/cifuzz-base/Dockerfile (hunk 18 above)
Comment (mpherman2) [reply infra/cifuzz/cifuzz-base/Dockerfile]: Done

File: infra/cifuzz/cifuzz_combined_entrypoint.py (hunk 19 above)
Comment (mpherman2) [reply infra/cifuzz/cifuzz_combined_entrypoint.py]: Done

File: infra/cifuzz/cifuzz_combined_entrypoint.py (hunk 20 above)
Comment (mpherman2) [reply infra/cifuzz/cifuzz_combined_entrypoint.py]: Done

**Review by jonathanmetzman — APPROVED** [review]
LGTM

**Pull request merged**